ipn/ipnlocal/local.go

```diff
@@ -6203,3 +6205,164 @@ func mayDeref[T any](p *T) (v T) {
+	candidateDistanceMap := make(map[float64][]tailcfg.NodeView, len(candidates))
+	preferredDerp := netMap.DERPMap.Regions[lastReport.PreferredDERP]
+	regions := make([]int, 0, len(netMap.DERPMap.Regions))
+	var minDistance float64 = math.MaxFloat64
+	if preferredDerp == nil {
+		return apitype.ExitNodeSuggestionResponse{ID: preferredExitNodeID, Name: "", Location: tailcfg.Location{}}, ErrNoPreferredDerp
+	}
+	for _, candidate := range candidates {
+		if candidate.Valid() {
+			if candidate.DERP() != "" {
+				ipp, err := netip.ParseAddrPort(candidate.DERP())
+				if err != nil {
+					continue
+				}
+				if ipp.Addr() == tailcfg.DerpMagicIPAddr {
+					regionID := int(ipp.Port())
+					if !slices.Contains(regions, regionID) {
+						regions = append(regions, regionID)
+					}
+					candidatesByRegion[regionID] = append(candidatesByRegion[regionID], candidate)
+				}
+			} else {
+				if candidate.Hostinfo().Location().View().Valid() {
+					distance := longLatDistance(preferredDerp.Latitude, preferredDerp.Longitude, candidate.Hostinfo().Location().Latitude, candidate.Hostinfo().Location().Longitude)
+					if distance < minDistance {
+						minDistance = distance
+					}
+					candidateDistanceMap[distance] = append(candidateDistanceMap[distance], candidate)
+				}
+			}
+		}
+	}
+	minRegion := minLatencyDERPRegion(regions, lastReport)
+	if candidatesByRegion[minRegion] != nil {
+		chosen := pickDERPNode(candidatesByRegion[minRegion], seed)
+		preferredExitNodeID = chosen.StableID()
+		preferredExitNodeName = chosen.Name()
+		hi := chosen.Hostinfo()
+		if hi.Valid() && hi.Location().View().Valid() {
+			preferredExitNodeLocation = *hi.Location()
+		}
+	} else {
+		const minDistanceEpsilon = 100000 // pick mullvad nodes that are at most 100km away from the minimum distance found.
+		var pickFrom []tailcfg.NodeView
+		for candidateDistance, candidate := range candidateDistanceMap {
+			if candidateDistance <= minDistance+minDistanceEpsilon {
+				pickFrom = append(pickFrom, candidate...)
+			}
+		}
+		chosen := pickWeighted(pickFrom)
+		if chosen.Valid() {
+			preferredExitNodeID = chosen.StableID()
+			preferredExitNodeName = chosen.Name()
+			preferredExitNodeLocation = *chosen.Hostinfo().Location()
+		}
+	}
+	if preferredExitNodeID.IsZero() {
+		return apitype.ExitNodeSuggestionResponse{ID: preferredExitNodeID, Name: "", Location: preferredExitNodeLocation}, nil
+	}
+	return apitype.ExitNodeSuggestionResponse{ID: preferredExitNodeID, Name: preferredExitNodeName, Location: preferredExitNodeLocation}, nil
+}
+
+// pickWeighted chooses the node with highest priority given a list of mullvad nodes.
+func pickWeighted(candidates []tailcfg.NodeView) tailcfg.NodeView {
+	maxWeight := 0
+	var chosenCandidate tailcfg.NodeView
+	for _, candidate := range candidates {
+		if loc := candidate.Hostinfo().Location(); loc != nil && loc.Priority > maxWeight {
+			maxWeight = loc.Priority
+			chosenCandidate = candidate
+		}
```

Comment:
This is short so it's OK as-is, but might be easier to read like this

```suggestion
		hi := candidate.HostInfo()
		if !hi.Valid() {
			continue
		}
		loc := hi.Location()
		if loc == nil || loc.Priority <= maxWeight {
			continue
		}
		maxWeight = loc.Priority
		chosenCandidate = candidate
```